### PR TITLE
Kv parity issues 94, 95, 96

### PR DESCRIPTION
**Pull request by scottf**
https://github.com/nats-io/nats-architecture-and-design/issues/94 Items 1 and 3
https://github.com/nats-io/nats-architecture-and-design/issues/95
https://github.com/nats-io/nats-architecture-and-design/issues/96

* 94: v2.7.2+, the KV stream need to be created with the DiscardNew
* 94: use Pending+Delivered.Consumer from the subscription to determine if end of data marker should be sent
* 95: use of prefix as defined in https://github.com/nats-io/nats-architecture-and-design/blob/main/adr/ADR-19.md
* 96 PurgeDeletes should keep "recent" markers

### CODE REVIEW

**Review by aricart — CHANGES_REQUESTED** [review]

File: src/main/java/io/nats/client/impl/NatsKeyValueManagement.java [hunk 1]
```diff
@@ -39,7 +39,11 @@
      */
     @Override
     public KeyValueStatus create(KeyValueConfiguration config) throws IOException, JetStreamApiException {
-        return new KeyValueStatus(jsm.addStream(config.getBackingConfig()));
+        StreamConfiguration sc = config.getBackingConfig();
+        if ( jsm.conn.getServerInfo().isOlderThanVersion("2.7.2") ) {
+            sc = StreamConfiguration.builder(sc).discardPolicy(null).build(); // null discard policy will use default
```
[the remark this answers is not in the document]
Comment (scottf) [reply src/main/java/io/nats/client/impl/NatsKeyValueManagement.java:44]: The default KeyValueConfiguration sets the backing stream config policy to Discard.New. This code is where the stream create is done and it's at this point that I check the server version. If the server version is old, the policy setting is cleared which means it gets the default of Old.

File: src/main/java/io/nats/client/KeyValue.java [hunk 2]
```diff
@@ -196,7 +193,8 @@
     List<KeyValueEntry> history(String key) throws IOException, JetStreamApiException, InterruptedException;
 
     /**
-     * Remove history from all keys that currently are deleted or purged.
+     * Remove history from all keys that currently are deleted or purged
+     * with using a default KeyValuePurgeOptions
      * THIS IS A BETA FEATURE AND SUBJECT TO CHANGE
```
Comment (ColinSullivan1) [inline src/main/java/io/nats/client/KeyValue.java:198]: Can remove this if this makes the KV API GA.  Alternatively, that could be another PR - up to you.

File: src/main/java/io/nats/client/api/KeyValuePurgeOptions.java [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package io.nats.client.api;
+
+import java.time.Duration;
+
+public class KeyValuePurgeOptions {
```
Comment (ColinSullivan1) [inline src/main/java/io/nats/client/api/KeyValuePurgeOptions.java:18]: Does this need doc?

File: src/main/java/io/nats/client/impl/NatsKeyValue.java [hunk 4]
```diff
@@ -35,32 +37,37 @@
...
-    private final String publishKeyPrefix;
+    private final String rawKeyPrefix;
+    private final String pubSubKeyPrefix;
 
     NatsKeyValue(NatsConnection connection, String bucketName, KeyValueOptions kvo) throws IOException {
         this.bucketName = Validator.validateKvBucketNameRequired(bucketName);
         streamName = toStreamName(bucketName);
         streamSubject = toStreamSubject(bucketName);
-        defaultKeyPrefix = toKeyPrefix(bucketName);
+        rawKeyPrefix = toKeyPrefix(bucketName);
```
Comment (ColinSullivan1) [inline src/main/java/io/nats/client/impl/NatsKeyValue.java:47]: Nitpick...  Could simplify a bit and remove rawKeyPrefix, just assign to pubSubKeyPrefix and reassign if necessary.

File: src/main/java/io/nats/client/impl/NatsKeyValue.java [hunk 5]
```diff
@@ -241,16 +244,53 @@ public NatsKeyValueWatchSubscription watchAll(KeyValueWatcher watcher, KeyValueW
...
+     */
+    @Override
+    public void purgeDeletes(KeyValuePurgeOptions options) throws IOException, JetStreamApiException, InterruptedException {
+        long dmThresh = options == null
+            ? KeyValuePurgeOptions.DEFAULT_THRESHOLD_MILLIS
+            : options.getDeleteMarkersThresholdMillis();
+
+        ZonedDateTime limit;
+        if (dmThresh < 0) {
+            limit = DateTimeUtils.fromNow(600000); // long enough in the future to clear all
```
Comment (ColinSullivan1) [inline src/main/java/io/nats/client/impl/NatsKeyValue.java:261]: May want to assign to a private constant...

Comment (ColinSullivan1) [reply src/main/java/io/nats/client/impl/NatsKeyValue.java:261]: Or configurable?

**Review by ColinSullivan1 — APPROVED** [review]
LGTM, have a few suggestions...

File: src/main/java/io/nats/client/api/KeyValuePurgeOptions.java (hunk 3 above)
Comment (scottf) [reply src/main/java/io/nats/client/api/KeyValuePurgeOptions.java:18]: yes, will do

File: src/main/java/io/nats/client/impl/NatsKeyValue.java (hunk 4 above)
Comment (scottf) [reply src/main/java/io/nats/client/impl/NatsKeyValue.java:47]: rawKeyPrefix is the base. pubSubKeyPrefix varies depending on whether or not there is a JetStream prefix
